src/broadcast/broadcast.rs

```diff
@@ -42,7 +42,7 @@ pub struct Broadcast<N> {
     /// /// Estimate as to how many nodes we think are faulty.
     fault_estimate: usize,
     /// The hashes and proofs we have received via `Echo` and `EchoHash` messages, by sender ID.
-    echos: BTreeMap<N, (Digest, Option<Proof<Vec<u8>>>)>,
+    echos: BTreeMap<N, (Digest, EchoMessage)>,
```

Comment:
Actually, since the `Proof` already contains the root hash, the additional `Digest` is redundant:
```rust
enum EchoContent {
    Hash(Digest),
    Full(Proof<Vec<u8>>>),
}
// …
    echos: BTreeMap<N, EchoContent>,
```
(Maybe `EchoContent` could have a `hash` method for convenience.)